<?xml version="1.0"?>

<robot name="transporter">
  <include filename="$(find transporter)/urdf/kinect.xacro" />
  <property name="M_PI" value="3.14159"/>
  <material name="Blue">
    <color rgba="0.1 0.1 0.8 1.0"/>
  </material>
  <material name="Green">
    <color rgba="0.1 0.6 0.0 0.2"/>
  </material>
  <material name="Black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <material name="Red">
    <color rgba="1.0 0.0 0.0 1.0"/>
  </material>

  <!-- base_footprint is a fictitious link(frame) that is on the ground right below base_link origin, navigation stack depends on this frame -->
  <link name="base_footprint">
    <inertial>
      <mass value="0.0001" />
      <inertia ixx="0.0001" ixy="0.0" ixz="0.0"
              iyy="0.0001" iyz="0.0" izz="0.0001" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.001 0.001 0.001" />
      </geometry>
      <material name="Green" />
    </visual>
    <collision>
      <origin xyz="0 0 0.118" rpy="0 0 0" />
      <geometry>
       <box size="0.001 0.001 0.001" />
      </geometry>
    </collision>
  </link> <!-- base_footprint -->

  <joint name="base_footprint_joint" type="fixed">
    <origin xyz="0 0 0.118" rpy="0 0 0" />        
    <parent link="base_footprint"/>
    <child link="base_link" />
  </joint>

  <link name="base_link">
    <inertial>
      <mass value="1" />
      <origin xyz="0 0 0" />
      <inertia ixx="1.0" ixy="0.0" ixz="0.0"
				iyy="1.0" iyz="0.0" izz="1.0" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder length="0.002" radius="0.200"/>
      </geometry>
      <material name="Blue"/>
    </visual>
    <collision>
      <origin xyz="0 0 0.059" rpy="0 0 0" />
      <geometry>
        <cylinder length="0.12" radius="0.300"/>
      </geometry>
    </collision>
  </link> <!-- base_link -->

  <joint name="wheel_axis_joint" type="fixed">
    <origin xyz="-0.1 0 -0.035" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="wheel_axis_link"/>
  </joint>

  <link name="wheel_axis_link">
    <inertial>
      <mass value="10" />
      <origin xyz="0 0 0" />
      <inertia ixx="1.0" ixy="0.0" ixz="0.0"
				iyy="1.0" iyz="0.0" izz="1.0" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.2 0.4 0.068" />
      </geometry>
      <material name="Blue" />
    </visual>
    <collision>
      <geometry>
       <box size="0.3 0.6 0.12" />
      </geometry>
    </collision>
  </link> <!-- wheel_axis_link -->

  <joint name="left_wheel_joint" type="continuous">
    <origin xyz="0 0.25 0.0" rpy="0 0 0"/>
    <parent link="wheel_axis_link"/>
    <child link="left_wheel_link"/>
    <axis xyz="0 1 0"/>
  </joint>

  <link name="left_wheel_link">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="0.1" />
      <inertia ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0" izz="1.0" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 ${M_PI/2} ${M_PI/2}" />
      <geometry>
        <cylinder radius="0.085" length = "0.10"/>
      </geometry>
      <material name="Black" />
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 ${M_PI/2} ${M_PI/2}" />
      <geometry>
        <cylinder radius="0.09" length = "0.2"/>
      </geometry>
    </collision>
  </link> <!-- left_wheel_link -->

  <joint name="right_wheel_joint" type="continuous">
    <origin xyz="0 -0.25 0.0" rpy="0 0 0"/>
    <parent link="wheel_axis_link"/>
    <child link="right_wheel_link"/>
    <axis xyz="0 1 0"/>
  </joint>

  <link name="right_wheel_link">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="0.1" />
      <inertia ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0" izz="1.0" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 ${M_PI/2} ${M_PI/2}" />
      <geometry>
        <cylinder radius="0.085" length = "0.10"/>
      </geometry>
      <material name="Black" />
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 ${M_PI/2} ${M_PI/2}" />
      <geometry>
        <cylinder radius="0.09" length = "0.2"/>
      </geometry>
    </collision>
  </link> <!-- right_wheel_link -->

  <!-- fixed because there's no transmission -->
  <joint name="front_castor_joint" type="fixed">
    <origin xyz="0.132 0 -0.059" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="front_wheel_link"/>
    <axis xyz="0 1 0"/>
  </joint>

  <link name="front_wheel_link">
    <inertial>
      <origin xyz="0 0 0" />
      <mass value="0.1" />
      <inertia ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0" izz="1.0" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 ${M_PI/2} ${M_PI/2}"/>
      <geometry>
        <cylinder radius="0.059" length = "0.05"/>
      </geometry>
      <material name="Green"/>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 ${M_PI/2} ${M_PI/2}"/>
      <geometry>
        <cylinder radius="0.059" length = "0.05"/>
      </geometry>
    </collision>
  </link> <!-- front_wheel_link -->

  <joint name="layer2_joint" type="fixed">
    <origin xyz="0 0 0.031" rpy="0 0 0" />
    <parent link="base_link"/>
    <child link="layer2_link" />
  </joint>

  <link name="layer2_link">
    <inertial>
      <mass value="0.1" />
      <origin xyz="0 0 0" />
      <inertia ixx="1.0" ixy="0.0" ixz="0.0"
        iyy="1.0" iyz="0.0" izz="1.0" />
    </inertial>
    <visual>
      <origin xyz=" 0 0 0 " rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.200" length = "0.035"/>
      </geometry>
	  <material name="Green"/>
    </visual>
    <collision>
      <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
      <geometry>
        <box size="0.450 0.450 0.100"/>
      </geometry>
    </collision>
  </link> <!-- layer2_link -->

  <!-- kinect parent="layer2_link" x_loc="0.0" y_loc="0.0" z_loc="0.15" rr="0" rp="0" ry="0"/ -->

  <joint name="gyro_joint" type="fixed">
    <origin xyz="0 0 0.04" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="gyro_link" />
  </joint>

  <link name="gyro_link">
    <visual>
      <geometry>
        <box size="0.064 0.064 0.026" />
      </geometry>
      <material name="Black" />
    </visual>
    <inertial>
      <mass value="0.075" />
      <origin xyz="0 0 0" />
      <inertia ixx="0.1" ixy="0.0" ixz="0.0"
        iyy="0.1" iyz="0.0" izz="0.1" />
    </inertial>
  </link>

  <joint name="laser_joint" type="fixed">
    <origin xyz="0.14 0 0.125" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="laser" />
  </joint>

  <link name="laser">
    <visual>
      <geometry>
        <box size="0.02 0.035 0.002" />
      </geometry>
      <material name="Red" />
    </visual>
    <inertial>
      <mass value="0.001" />
      <origin xyz="0 0 0" />
      <inertia ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.0001" />
    </inertial>
  </link>

</robot>
